src/Syntax_Server/State.re

```diff
@@ -97,7 +101,50 @@ let doPendingWork = state => {
       state.visibleBuffers,
     );
 
-  {...state, highlightsMap};
+  // Apply any newly discovered keywords
+  let keywordMap =
+    List.fold_left(
+      (acc, bufferId) => {
+        let discoveredKeywords =
+          state.highlightsMap
+          |> IntMap.find_opt(bufferId)
+          |> Option.map(NativeSyntaxHighlights.getDiscoveredKeywords);
+
+        let scope = state.bufferIdToScope |> IntMap.find_opt(bufferId);
+
+        Option.map2(
+          (keywords: list((int, list(string))), scope: string) => {
+            List.fold_left(
+              (acc, curr) => {
+                let (line, words) = curr;
+                KeywordMap.set(
+                  ~bufferId,
+                  ~scope,
+                  // TODO: Get lines from keywords
+                  ~line,
+                  ~words,
+                  acc,
+                );
+              },
+              acc,
+              keywords,
+            )
+          },
+          discoveredKeywords,
+          scope,
+        )
+        |> Option.value(~default=acc);
+      },
+      state.keywordMap,
+      state.visibleBuffers,
+    );
```

Comment:
It's usually better to just name the function than have a large anonymous function effectively described by a comment.